### PR TITLE
Thread local: wait on main thread until worker's slots are released

### LINKED ISSUE #32427: Envoy occasionally segfaults from v1.27 in ProxyFilterConfig::onClusterAddOrUpdate
**Title**: Envoy occasionally segfaults from v1.27 in ProxyFilterConfig::onClusterAddOrUpdate

**Description**:
  Since we upgraded envoy version to v1.27.2, we see Envoy crashes a few times a day inside ProxyFilterConfig::onClusterAddOrUpdate function in extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc.
  
  The function was added to v1.27, (added by [#26420](https://github.com/envoyproxy/envoy/pull/26420)), and we verified that it also happens with envoy v1.28.1. After we rolled Envoy to v1.26.6, we do not see the issue.

  We use envoy binaries from the [prebuilt docker images](https://www.envoyproxy.io/docs/envoy/latest/start/install#pre-built-envoy-docker-images). We use the `debug` variant  which contains debug symbols. (e.g. [debug-v1.27.2](https://hub.docker.com/layers/envoyproxy/envoy/debug-v1.27.2/images/sha256-880a6ca106553d5ed1ca4a0f2459fb9fed591367526ff896aca7737e795c8372?context=explore)]

 We use envoy proxy as an edge proxy, and we are running multiple envoy proxies running that fetch configuration using ADS from the same management server written in go-control-plane(v0.11.1). Since a single envoy proxy crashes randomly, and the crash happens on a different envoy proxy each time, it appears that this happens when a certain race condition is met. 

 The segfault happens since `tls_slot_` has a null value. (see troubleshooting logs below). However, we are not sure when the value can be null. 

**Repro steps**:

 Unfortunately, we were not able to deterministically reproduce the issue. We have multiple services that use dynamic forwarding proxy to forward requests to upstream endpoints, and we frequently update the configurations. (updates can happen multiple times a minute)

**Call Stack**:

```
* thread #1, name = 'envoy-debug-1.2', stop reason = signal SIGSEGV
  * frame #0: 0x000056241c27a030 envoy-debug-1.27.2`Envoy::Extensions::HttpFilters::DynamicForwardProxy::ProxyFilterConfig::onClusterAddOrUpdate(Envoy::Upstream::ThreadLocalCluster&) at thread_local.h:48:29
    frame #1: 0x000056241d7cea51 envoy-debug-1.27.2`std::__1::__function::__func<Envoy::Upstream::ClusterManagerImpl::postThreadLocalClusterUpdate(Envoy::Upstream::ClusterManagerCluster&, Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterUpdateParams&&)::$_19, std::__1::allocator<Envoy::Upstream::ClusterManagerImpl::postThreadLocalClusterUpdate(Envoy::Upstream::ClusterManagerCluster&, Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterUpdateParams&&)::$_19>, void (Envoy::OptRef<Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterManagerImpl>)>::operator()(Envoy::OptRef<Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterManagerImpl>&&) at cluster_manager_impl.cc:1152:24
    frame #2: 0x000056241d7cac08 envoy-debug-1.27.2`decltype(static_cast<Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterManagerImpl>(fp)(static_cast<Envoy::ThreadLocal::TypedSlot<Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterManagerImpl>::makeSlotUpdateCb(std::__1::function<void (Envoy::OptRef<Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterManagerImpl>)>)::'lambda'(std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject>)&>(fp0)...)) std::__1::__invoke<Envoy::ThreadLocal::TypedSlot<Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterManagerImpl>::makeSlotUpdateCb(std::__1::function<void (Envoy::OptRef<Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterManagerImpl>)>)::'lambda'(std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject>)&, std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject> >(Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterManagerImpl&&, Envoy::ThreadLocal::TypedSlot<Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterManagerImpl>::makeSlotUpdateCb(std::__1::function<void (Envoy::OptRef<Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterManagerImpl>)>)::'lambda'(std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject>)&...) at function.h:507:16
    frame #3: 0x000056241d762c94 envoy-debug-1.27.2`std::__1::__function::__func<Envoy::ThreadLocal::InstanceImpl::SlotImpl::dataCallback(std::__1::function<void (std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject>)> const&)::$_1, std::__1::allocator<Envoy::ThreadLocal::InstanceImpl::SlotImpl::dataCallback(std::__1::function<void (std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject>)> const&)::$_1>, void ()>::operator()() at function.h:507:16
    frame #4: 0x000056241deb6ea7 envoy-debug-1.27.2`Envoy::Event::DispatcherImpl::runPostCallbacks() at any_invocable.h:859:1
    frame #5: 0x000056241e49c81a envoy-debug-1.27.2`event_process_active_single_queue + 666
    frame #6: 0x000056241e49b431 envoy-debug-1.27.2`event_base_loop + 1953
    frame #7: 0x000056241d7861fb envoy-debug-1.27.2`Envoy::Server::WorkerImpl::threadRoutine(Envoy::Server::GuardDog&, std::__1::function<void ()> const&) at worker_impl.cc:148:16
    frame #8: 0x000056241e4a3c53 envoy-debug-1.27.2`Envoy::Thread::ThreadImplPosix::ThreadImplPosix(std::__1::function<void ()>, std::__1::optional<Envoy::Thread::Options> const&)::'lambda'(void*)::__invoke(void*) at function.h:507:16
    frame #9: 0x00007fb5a2e0aea7 libpthread.so.0`start_thread(arg=<unavailable>) at pthread_create.c:477:8
    frame #10: 0x00007fb5a2d2aa2f libc.so.6`__clone at clone.S:95
```
Below are outputs from lldb we obtained using the core dump file from a crash

disassembled output of frame #0
```
(lldb) disassemble
envoy-debug-1.27.2`Envoy::Extensions::HttpFilters::DynamicForwardProxy::ProxyFilterConfig::onClusterAddOrUpdate:
...
    0x55c2f69c4027 <+375>:  callq  0x1701130                 ; void spdlog::logger::log_<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>(spdlog::source_loc, spdlog::level::level_enum, fmt::v9::basic_string_view<char>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
    0x55c2f69c402c <+380>:  movq   0x48(%r14), %rsi
->  0x55c2f69c4030 <+384>:  movq   (%rsi), %rax
```

we can see the value of `rsi` register is 0, which caused `movq` instruction to segfault
```
(lldb) register read $rsi
     rsi = 0x0000000000000000

```

`rsi` has the value at address `%r14 + 0x48`
```
(lldb) memory read --size 8 -fx -c 2 $r14+0x48
0x26d94193c3e0: 0x0000000000000000 0x0000000000001388
```

%r14 points to the `this` object, and %r14 + 0x48 points to `this->tls_slot` (the next value 0x1388(=5000) is the value of the next field "cluster_init_timeout_" )

### PR DESCRIPTION
**Pull request by cpakulski**
Commit Message:
Thread local: wait on main thread until worker's slots are released
Additional Description:
This is an alternative way of solving race-condition reported in #32427. The first attempts was in PR #32692, which also contains more info about the race condition.
I believe that this PR solves the problem in more generic way.
The problem is that very often thread local objects stored in workers' threads refer to objects owned by the main thread. If the main thread does not wait until workers' thread release their slots and stop referring to those objects, the main thread may release those objects while workers threads may still have references to them.
Risk Level: Low
Testing: Modified unit tests to reflect new logic.
Docs Changes: No
Release Notes: No
Platform Specific Features: No
Fixes #32427

### CODE REVIEW

**Comment by cpakulski** [comment]
@alyssawilk This is related to PR #32692, which you reviewed. I believe that this PR is much cleaner way of solving the problem. Would you mind reviewing it or pointing me to the right person? Thanks!

**Comment by alyssawilk** [comment]
looks like @kyessenov was assigned first pass.  @kyessenov ping?

**Comment by kyessenov** [comment]
@alyssawilk Reviewed. I am not happy about the blocking solution, but I haven't reviewed DFP code in depth before, so perhaps, you have a better understanding of the issue.

**Comment by cpakulski** [comment]
Thanks for reviewing @kyessenov . 

> The skew between workers can be significant, I've seen it close to seconds, which means a main may be potentially blocked for a second and not do useful work.

I did not know that it may go as much out-of-sync. 

Let me experiment with wrapping, as you suggested, in shared pointer. That solution was also suggested in https://github.com/envoyproxy/envoy/pull/32692#pullrequestreview-1923107947. I will try to do it at low level, in generic form, but if it does not work, I will change signature for registering callbacks function to take shared ptr.

File: source/common/thread_local/thread_local_impl.cc [hunk 1]
```diff
@@ -136,15 +139,43 @@ void InstanceImpl::removeSlot(uint32_t slot) {
...
     if (slot < thread_local_data_.data_.size()) {
       thread_local_data_.data_[slot] = nullptr;
     }
+    if (auto counter = counter_weak.lock()) {
+      counter->DecrementCount();
+    }
   });
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.thread_local_synchronize_slots_deletion")) {
+    counter->Wait();
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/common/thread_local/thread_local_impl.cc:177]: ugh so sorry, I think I gave you poor advice by not actually understanding the underlying problem.

Ok here's my new thought, but I'd appreciate a chime in from @mattklein123 in case there's some clever workaround because my last suggestion wasn't great :-/

Tldr for Matt please correct if not accurate @cpakulski  from the linked issue and first attempt, each DFP filter registers with the global CM for cluster updates.  when the DFP filter goes away, there's a race where the ClusterUpdateCallbacksHandlePtr is deleted but the the main thread is still sending updates.

I now think the correct thing to do is add a big warning note about how thread local slots will be deleted and potential race conditions.  Then make ClusterUpdateCallbacksHandlePtr have a release() function which makes up-calls stop immediately, then either make it a shared pointer or hand ownership to a post to the main thread dispatcher so it's not deleted until after the main thread stops referencing it.
/wait-any

[the remark this answers is not in the document]
Comment (cpakulski) [reply source/common/thread_local/thread_local_impl.cc:177]: > ugh so sorry, I think I gave you poor advice by not actually understanding the underlying problem.

@alyssawilk, you actually gave the correct advice to try shared pointer: https://github.com/envoyproxy/envoy/pull/32692#pullrequestreview-1923107947. 
That would however require some restructuring of DFP filter and maybe some other filters which register callbacks. I thought that before trying that, maybe the problem can be fixed at thread local level. My next step is to try intermediate shared pointer, which can be used to wrap callbacks embedded in DFP filter. If it works, it will also be generic solution and will not require DFP filter changes. I am working on that. If it does not work or is ugly, I will try to restructure DFP filter. 

> Tldr for Matt please correct if not accurate @cpakulski from the linked issue and first attempt, each DFP filter registers with the global CM for cluster updates. when the DFP filter goes away, there's a race where the ClusterUpdateCallbacksHandlePtr is deleted but the the main thread is still sending updates.

That is close description, except one small detail:

each DFP filter registers with the global CM for cluster updates. when the DFP filter goes away, there's a race where one _of the worker threads has not-finished yet (scheduled or is currently executing) job which references callbacks which were deleted by DFP on main thread_.

[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/thread_local/thread_local_impl.cc:177]: Sorry I would need to sit down with the code and look at it all to have good advice. I can do that if needed. In general I agree with @kyessenov that in the past we have extended the lifetimes of things as needed and/or made sure that destruction happens on the main thread for cases like this.

[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/common/thread_local/thread_local_impl.cc:177]: oh phew.  yeah I think having a cross-thread lock is going to be more problematic than shared pointer, which is also inline with Matt's advice about extending lifetime so I think we want to go that route.

**Comment by alyssawilk** [comment]
/wait for now in that case.  Fingers crossed!

**Comment by cpakulski** [comment]
Closing. Fixed in #33303.
